src/codegen/register_allocation.c

```diff
@@ -557,15 +693,19 @@ void coalesce(IRFunction *f, const MachineDescription *desc, IRInstructions *ins
 
         /// If we get here, then we know that all PHI node that use to
         /// can be precoloured with the same register as from.
-        foreach_ptr (IRInstruction *, phi, phis) { phi->result = from->result; }
+        foreach_ptr (IRInstruction *, phi, phis) phi->result = from->result;
+        DEBUG("All PHI nodes that use `to` can be precouloured with the same register as from\n");
         goto eliminate;
       }
 
+      /*
       /// Neither from nor to are precoloured, and they do not interfere.
       /// Replace all uses of to with from, and eliminate the copy.
       if (!to->result && !from->result && !adjm(G->matrix, to->index, from->index)) {
```

Comment:
NOTE: I attempted to add ` && !check_register_interference(G->regmasks[to->index], from)` to this `if` condition, but it did not seem to fix anything with the coalescing; removing it does **sometimes** create sub-optimal programs with more copies than necessary, but the programs it creates are **correct**, which I feel is more important.